libaugrim/src/algorithm/two_phase_commit/participant_algorithm.rs

```diff
@@ -0,0 +1,373 @@
+// Copyright 2021-2022 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::marker::PhantomData;
+use std::time::Duration;
+
+use crate::algorithm::{Algorithm, Value};
+use crate::error::{AlgorithmError, InvalidStateError};
+use crate::process::Process;
+use crate::time::TimeSource;
+
+use super::ParticipantAction;
+use super::ParticipantActionAlarm;
+use super::ParticipantActionNotification;
+use super::ParticipantContext;
+use super::ParticipantEvent;
+use super::ParticipantMessage;
+use super::ParticipantState;
+use super::TwoPhaseCommitMessage;
+
+const DECISION_TIMEOUT_SECONDS: u64 = 30;
+
+pub struct ParticipantAlgorithm<P, V, TS>
+where
+    P: Process,
+    V: Value,
+    TS: TimeSource,
+{
+    _process_phantom: PhantomData<P>,
+    _value_phantom: PhantomData<V>,
+    time_source: TS,
+}
+
+impl<P, V, TS> ParticipantAlgorithm<P, V, TS>
+where
+    P: Process,
+    V: Value,
+    TS: TimeSource,
+{
+    pub fn new(time_source: TS) -> Self {
+        ParticipantAlgorithm {
+            _process_phantom: PhantomData,
+            _value_phantom: PhantomData,
+            time_source,
+        }
+    }
+}
+
+impl<P, V, TS> Algorithm<P> for ParticipantAlgorithm<P, V, TS>
+where
+    P: Process,
+    V: Value,
+    TS: TimeSource,
+{
+    type Event = ParticipantEvent<P, V>;
+    type Action = ParticipantAction<P, V, TS::Time>;
+    type Context = ParticipantContext<P, TS::Time>;
+
+    fn event(
+        &self,
+        event: Self::Event,
+        mut context: Self::Context,
+    ) -> Result<Vec<Self::Action>, AlgorithmError> {
+        match event {
+            ParticipantEvent::Alarm() => match context.state().clone() {
```

Comment:
Is this clone necessary?

Reply:
A clone is necessary because of the ParticipantAction::Update below.

I moved the clone to context, more local to where the update occurs, but while its clearer, it clones a larger amount.